### PR TITLE
Add reattachment of biological limbs

### PR DESCRIPTION
**Pull request by spookydonut**
## About The Pull Request
lets surgeons reattach original limbs to their owners

## Why It's Good For The Game
gives a reason to recover your missing arm and some challenging things for doctors to do

## Changelog
:cl:
add: Added the ability to reattach biological limbs
/:cl:

### CODE REVIEW

File: code/modules/organs/limb_objects.dm [hunk 1]
```diff
@@ -1,8 +1,31 @@
+#define LIMB_REATTACH_TIME 15 MINUTES
 
-obj/item/limb
+/obj/item/limb
 	icon = 'icons/mob/human_races/r_human.dmi'
-
-obj/item/limb/New(loc, mob/living/carbon/human/H)
+	var/amputation_time
+	var/mob/living/carbon/human/owner
```
Comment (Rohesie) [inline code/modules/organs/limb_objects.dm]: Can this be carbon? Monkeys and other things have limbs.

**Comment by Rohesie** [comment]
Ok, this is pretty amazing.

File: code/modules/organs/limb_objects.dm (hunk 1 above)
Comment (Ghommie) [reply code/modules/organs/limb_objects.dm]: As of now only humans seems to have limbs, monkeys don't.

Comment (Rohesie) [reply code/modules/organs/limb_objects.dm]: I see. It's something that we'll want in the future, though, most likely.

Comment (spookydonut) [reply code/modules/organs/limb_objects.dm]: when they have limbs this can be changed to carbon.

**Comment by spookydonut** [comment]
This is turning into a refactor of limbs/wounds, will atomise those away.

File: code/modules/mob/living/carbon/human/life/handle_organs.dm [hunk 2]
```diff
@@ -31,40 +31,45 @@
...
-		for(var/datum/limb/E in limbs)
-			if(!E)
-				continue
-			if(!E.need_process())
-				//bad_limbs -= E
-				continue
-			else
-				E.process()
+	//processing internal organs is pretty cheap, do that first.
+	for(var/datum/internal_organ/I in internal_organs)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/life/handle_organs.dm]:
```suggestion
	for(var/X in internal_organs)
		var/datum/internal_organ/I = X
```

File: code/modules/mob/living/carbon/human/life/handle_organs.dm [hunk 3]
```diff
@@ -31,40 +31,45 @@
...
+		I.process()
 
-				if (!lying && world.time - l_move_time < 15)
-					if(m_intent != MOVE_INTENT_WALK || pulledby) //Running around with fractured ribs won't do you any good; walking prevents worsening, unless you're being pulled around
-						if (E.is_broken() && E.internal_organs && prob(15))
-							var/datum/internal_organ/I = pick(E.internal_organs)
-							custom_pain("You feel broken bones moving in your [E.display_name]!", 1)
-							I.take_damage(rand(3,5))
+	for(var/datum/limb/E in limbs)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/life/handle_organs.dm]:
```suggestion
	for(var/X in limbs)
		var/datum/limb/E = X
```

File: code/modules/mob/living/carbon/human/life/handle_organs.dm [hunk 4]
```diff
@@ -31,40 +31,45 @@
...
+			if (!lying && world.time - l_move_time < 15)
+				if(m_intent != MOVE_INTENT_WALK || pulledby) //Running around with fractured ribs won't do you any good; walking prevents worsening, unless you're being pulled around
+					if (E.is_broken() && E.internal_organs && prob(15))
+						var/datum/internal_organ/I = pick(E.internal_organs)
+						custom_pain("You feel broken bones moving in your [E.display_name]!", 1)
+						I.take_damage(rand(3,5))
+
+				//Moving makes open wounds get infected much faster
+				if (E.wounds.len)
+					for(var/datum/wound/W in E.wounds)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/life/handle_organs.dm]:
```suggestion
					for(var/Y in E.wounds)
						var/datum/wound/W = Y
```

File: code/modules/mob/living/carbon/human/life/handle_organs.dm [hunk 5]
```diff
@@ -31,40 +31,45 @@
...
+		if(!E.need_process())
+			continue
+		else
+			E.process()
+
+			if (!lying && world.time - l_move_time < 15)
+				if(m_intent != MOVE_INTENT_WALK || pulledby) //Running around with fractured ribs won't do you any good; walking prevents worsening, unless you're being pulled around
+					if (E.is_broken() && E.internal_organs && prob(15))
+						var/datum/internal_organ/I = pick(E.internal_organs)
+						custom_pain("You feel broken bones moving in your [E.display_name]!", 1)
```
Comment (Rohesie) [inline code/modules/mob/living/carbon/human/life/handle_organs.dm]: This is being done twice, once by the signal and once here, no?

File: code/modules/organs/wound.dm [hunk 6]
```diff
@@ -52,160 +47,191 @@
...
+	for(var/V in stages)
+		desc_list += V
+		damage_list += stages[V]
+
+	src.damage = damage
+
+	max_bleeding_stage = src.desc_list.len - max_bleeding_stage
+
+	// initialize with the appropriate stage
+	src.init_stage(damage)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	init_stage(damage)
```

File: code/modules/organs/wound.dm [hunk 7]
```diff
@@ -52,160 +47,191 @@
...
+
+/datum/wound/Destroy(force, ...)
+	limb.wounds -= src
+	return ..()
+
+// returns 1 if there's a next stage, 0 otherwise
+/datum/wound/proc/init_stage(var/initial_damage)
+	current_stage = stages.len
+
+	while(src.current_stage > 1 && src.damage_list[current_stage-1] <= initial_damage / src.amount)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	while(current_stage > 1 && damage_list[current_stage-1] <= initial_damage / amount)
```

File: code/modules/organs/wound.dm [hunk 8]
```diff
@@ -52,160 +47,191 @@
...
+		src.current_stage--
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
		current_stage--
```

File: code/modules/organs/wound.dm [hunk 9]
```diff
@@ -52,160 +47,191 @@
...
+
+	src.min_damage = damage_list[current_stage]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	min_damage = damage_list[current_stage]
```

File: code/modules/organs/wound.dm [hunk 10]
```diff
@@ -52,160 +47,191 @@
...
+	src.desc = desc_list[current_stage]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	desc = desc_list[current_stage]
```

File: code/modules/organs/wound.dm [hunk 11]
```diff
@@ -52,160 +47,191 @@
...
+
+// the amount of damage per wound
+/datum/wound/proc/wound_damage()
+	return src.damage / src.amount
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	return damage / amount
```

File: code/modules/organs/wound.dm [hunk 12]
```diff
@@ -52,160 +47,191 @@
...
+
+/datum/wound/proc/can_autoheal()
+	if(src.wound_damage() <= autoheal_cutoff)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	if(wound_damage() <= autoheal_cutoff)
```

File: code/modules/organs/wound.dm [hunk 13]
```diff
@@ -52,160 +47,191 @@
...
+	germ_level = 0
+
+/datum/wound/proc/clamp()
+	APPLY_WOUND_FLAG(WOUND_CLAMPED)
+
+/datum/wound/proc/salve()
+	APPLY_WOUND_FLAG(WOUND_SALVED)
+
+/datum/wound/proc/can_disinfect()
+	return (germ_level > 0 && germ_level < 50 && CHECK_BITFIELD(wound_flags, WOUND_SALVED))
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	switch(germ_level)
		if(1 to 50)
			return CHECK_BITFIELD(wound_flags, WOUND_SALVED)
```

File: code/modules/organs/wound.dm [hunk 14]
```diff
@@ -52,160 +47,191 @@
...
+// set @heals_internal to also heal internal organ damage
+/datum/wound/proc/heal_damage(amount, heals_internal = 0)
+	if(CHECK_BITFIELD(wound_flags, WOUND_INTERNAL) && !heals_internal)
+		// heal nothing
+		return amount
 
-		var/healed_damage = min(src.damage, amount)
-		amount -= healed_damage
-		src.damage -= healed_damage
+	var/healed_damage = min(src.damage, amount)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	var/healed_damage = min(damage, amount)
```

File: code/modules/organs/wound.dm [hunk 15]
```diff
@@ -52,160 +47,191 @@
...
+	amount -= healed_damage
+	src.damage -= healed_damage
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	damage -= healed_damage
```

File: code/modules/organs/wound.dm [hunk 16]
```diff
@@ -52,160 +47,191 @@
...
 
-		while(src.wound_damage() < damage_list[current_stage] && current_stage < src.desc_list.len)
-			current_stage++
-		desc = desc_list[current_stage]
-		src.min_damage = damage_list[current_stage]
+	while(src.wound_damage() < damage_list[current_stage] && current_stage < src.desc_list.len)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	while(wound_damage() < damage_list[current_stage] && current_stage < src.desc_list.len)
```

File: code/modules/organs/wound.dm [hunk 17]
```diff
@@ -52,160 +47,191 @@
...
+		current_stage++
+	desc = desc_list[current_stage]
+	src.min_damage = damage_list[current_stage]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	min_damage = damage_list[current_stage]
```

File: code/modules/organs/wound.dm [hunk 18]
```diff
@@ -52,160 +47,191 @@
...
-		src.damage += damage
-		bleed_timer += damage
+// opens the wound again
+/datum/wound/proc/open_wound(damage)
+	src.damage += damage
+	bleed_timer += damage
 
-		while(src.current_stage > 1 && src.damage_list[current_stage-1] <= src.damage / src.amount)
-			src.current_stage--
+	while(src.current_stage > 1 && src.damage_list[current_stage-1] <= src.damage / src.amount)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	while(current_stage > 1 && damage_list[current_stage-1] <= src.damage / amount)
```

File: code/modules/organs/wound.dm [hunk 19]
```diff
@@ -52,160 +47,191 @@
...
+		src.current_stage--
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
		current_stage--
```

File: code/modules/organs/wound.dm [hunk 20]
```diff
@@ -52,160 +47,191 @@
...
 
-		src.desc = desc_list[current_stage]
-		src.min_damage = damage_list[current_stage]
+	src.desc = desc_list[current_stage]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	desc = desc_list[current_stage]
```

File: code/modules/organs/wound.dm [hunk 21]
```diff
@@ -52,160 +47,191 @@
...
+	src.min_damage = damage_list[current_stage]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	min_damage = damage_list[current_stage]
```

File: code/modules/organs/wound.dm [hunk 22]
```diff
@@ -52,160 +47,191 @@
...
-			return 0	//incompatible damage types
+// returns whether this wound can absorb the given amount of damage.
+// this will prevent large amounts of damage being trapped in less severe wound types
+/datum/wound/proc/can_worsen(damage_type, damage)
+	if (src.damage_type != damage_type)
+		return 0	//incompatible damage types
 
-		if (src.amount > 1)
-			return 0
+	if (src.amount > 1)
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	if (amount > 1)
```

File: code/modules/organs/wound.dm [hunk 23]
```diff
@@ -52,160 +47,191 @@
...
-		//with 1.5*, a shallow cut will be able to carry at most 30 damage,
-		//37.5 for a deep cut
-		//52.5 for a flesh wound, etc.
-		var/max_wound_damage = 1.5*src.damage_list[1]
-		if (src.damage + damage > max_wound_damage)
-			return 0
+	//with 1.5*, a shallow cut will be able to carry at most 30 damage,
+	//37.5 for a deep cut
+	//52.5 for a flesh wound, etc.
+	var/max_wound_damage = 1.5*src.damage_list[1]
```
Comment (Rohesie) [inline code/modules/organs/wound.dm]:
```suggestion
	var/max_wound_damage = 1.5 * damage_list[1]
```

File: code/modules/surgery/limb_reattach.dm [hunk 24]
```diff
@@ -0,0 +1,194 @@
+//scalpel -> retract -> place limb -> hemo -> fix-o-vein  -> bone gel -> bone setter ->  ATK -> cautery
+
+/datum/surgery_step/limb_reattach
+	priority = 1
+	can_infect = 0
```
Comment (Rohesie) [inline code/modules/surgery/limb_reattach.dm]:
```suggestion
	can_infect = FALSE
```

File: code/modules/surgery/limb_reattach.dm [hunk 25]
```diff
@@ -0,0 +1,194 @@
+//scalpel -> retract -> place limb -> hemo -> fix-o-vein  -> bone gel -> bone setter ->  ATK -> cautery
+
+/datum/surgery_step/limb_reattach
+	priority = 1
+	can_infect = 0
+	var/reattach_step = 0
+
+/datum/surgery_step/limb_reattach/can_use(mob/living/user, mob/living/carbon/human/target, target_zone, obj/item/tool, datum/limb/affected, checks_only)
+	if(!affected)
+		to_chat(world, "cant use9")
```
Comment (Rohesie) [inline code/modules/surgery/limb_reattach.dm]: Only noticed now this is WIP. RIP.

**Comment by spookydonut** [comment]
This is still a draft PR.

**Pull request closed without merging**